source-kafka/src/kafka.rs

```diff
@@ -92,6 +92,29 @@ pub fn subscribe(
     Ok(topic_partition_list)
 }
 
+pub fn high_watermarks(
+    consumer: &BaseConsumer,
+    topics: &state::TopicSet,
+) -> Result<state::TopicSet, Error> {
+    let mut watermarks = state::TopicSet::default();
+
+    for topic in topics.0.iter() {
+        let (_low, high) = consumer
+            .fetch_watermarks(&topic.name, topic.partition, KAFKA_TIMEOUT)
+            .map_err(Error::Metadata)?;
+        watermarks.add_new(state::Topic::new(
+            &topic.name,
+            topic.partition,
+            // We subtract 1 from the high watermark since it is zero-indexed.
+            // If we try to read up *through* `high`, we'll potentially sit here
+            // forever waiting on the next message to come through.
+            state::Offset::UpThrough(high - 1),
```

Comment:
Okay, I dug into this a little further. First, this API is a bit silly. They don't even return `(low, high)` in a specific order. `librdkafka` interprets the higher value as `high` and the lower value as `low`. :roll_eyes: 

As far as the off-by-one potential here. I think the edgecase to watch out for here is `low == high`, rather than `low == high - 1` like we initially discussed. 
* `high` is going to be the offset of the _next_ message to be added to the partition.
* `low` is the offset of the first message available to be read in the partition.

A few examples:
- `low = 0,  high = 50` would mean we have 50 messages to read, Message 0 through Message 49.
- `low = 49, high = 50` would mean that we *can* read Message 49, so we should.
- `low = 50, high = 50` would mean that the next message is Message 50, but there aren't any messages to read until that one comes in.

Hilariously, if `low == high`, they don't even send 2 offsets back from the broker. :rofl: 

https://github.com/edenhill/librdkafka/blob/5aa9ea865b5a79cd7d345a6816b9e2b9c1abb903/src/rdkafka.c#L3512-L3514